lib/src/view/analysis/tree_view.dart

```diff
@@ -162,124 +131,654 @@ class _InlineTreeViewState extends ConsumerState<AnalysisTreeView> {
+        ),
+        if (children.isNotEmpty)
+          _IndentedSideLines(
+            children,
+            parent: sidelineNodes.last,
+            initialPath: UciPath.join(
+              initialPath,
+              UciPath.fromIds(sidelineNodes.map((node) => node.id)),
+            ),
+            params: params,
+            nesting: nesting + 1,
           ),
-        );
-      } else {
-        widgets.addAll(
-          _buildTreeWidget(
-            pgn,
-            options,
-            parent: parent,
-            nodes: [nodes[i]].lockUnsafe,
-            shouldShowAnnotations: shouldShowAnnotations,
-            shouldShowComments: shouldShowComments,
-            inMainline: false,
-            startMainline: false,
-            startSideline: true,
-            initialPath: initialPath,
-          ),
-        );
+      ],
+    );
+  }
+}
+
+class _IndentPainter extends CustomPainter {
+  const _IndentPainter({
+    required this.sideLineStartPositions,
+    required this.color,
+    required this.padding,
+  });
+
+  final List<Offset> sideLineStartPositions;
+
+  final Color color;
+
+  final double padding;
+
+  @override
+  void paint(Canvas canvas, Size size) {
+    if (sideLineStartPositions.isNotEmpty) {
+      final paint = Paint()
+        ..strokeWidth = 1.5
+        ..color = color
+        ..strokeCap = StrokeCap.round
+        ..style = PaintingStyle.stroke;
+
+      final origin = Offset(-padding, 0);
+
+      final path = Path()..moveTo(origin.dx, origin.dy);
+      path.lineTo(origin.dx, sideLineStartPositions.last.dy);
+      for (final position in sideLineStartPositions) {
+        path.moveTo(origin.dx, position.dy);
+        path.lineTo(origin.dx + padding / 2, position.dy);
       }
+      canvas.drawPath(path, paint);
     }
+  }
 
-    // add the children of the first child
-    widgets.addAll(
-      _buildTreeWidget(
-        pgn,
-        options,
-        parent: firstChild,
-        nodes: firstChild.children,
-        shouldShowAnnotations: shouldShowAnnotations,
-        shouldShowComments: shouldShowComments,
-        inMainline: inMainline,
-        startMainline: false,
-        startSideline: false,
-        initialPath: newPath,
-      ),
+  @override
+  bool shouldRepaint(_IndentPainter oldDelegate) {
+    return oldDelegate.sideLineStartPositions != sideLineStartPositions ||
+        oldDelegate.color != color;
+  }
+}
+
+class _IndentedSideLines extends StatefulWidget {
+  const _IndentedSideLines(
+    this.sideLines, {
+    required this.parent,
+    required this.initialPath,
+    required this.params,
+    required this.nesting,
+  });
+
+  final Iterable<ViewBranch> sideLines;
+
+  final ViewNode parent;
+
+  final UciPath initialPath;
+
+  final _PgnTreeViewParams params;
+
+  final int nesting;
+
+  @override
+  State<_IndentedSideLines> createState() => _IndentedSideLinesState();
+}
+
+class _IndentedSideLinesState extends State<_IndentedSideLines> {
+  late List<GlobalKey> _keys;
```

Comment:
In flutter we tend to avoid using `GlobalKey` unless having a good reason. Here I'm a little worried about this list of global keys.

I understand you need them to get the position of the sideline to get the proper offset. I think there is a way (no much harder to code I'd say) to do with without all the global keys. The indent width is a fixed one, so in theory you'd need to do sth like:
```dart
final indentOffset = initialOffset + Offset(nestingLevel * fixedIndentWidth, 0)`;
```
So you'd need only one global key to get the initial offset of the tree view, but not more.

Reply:
The problem is that this only works if the sidelines do not wrap. As soon as a sidelines wraps, we don't know the vertical positions of all following sidelines.

(Or maybe I did not understand your suggestions correctly)

Reply:
I suspected something like that, but I had to be sure to understand so pardon me for the question :)

As long as there's no performance issue, this list of global keys shouldn't cause any problem as the scope is limited.

Reply:
Yeah no problem, I went back and forth about this many times in my head as well, but in the end the problem always was "what if the line wraps though?"